src/Microsoft.AspNetCore.Sockets.Common/Internal/Formatters/ServerSentEventsMessageParser.cs

```diff
@@ -0,0 +1,193 @@
+                        {
+                            throw new FormatException("Expected the message prefix 'data: '");
+                        }
+
+                        //Slice away the 'data: '
+                        var newData = line.Slice(6).ToArray();
+
+                        start = lineEnd;
+                        _data.Add(newData);
+                        consumed = lineEnd;
+                        break;
+                    case InternalParseState.ReadEndOfMessage:
+                        if (ReadCursorOperations.Seek(start, end, out lineEnd, ByteLF) == -1)
+                        {
+                            // The message has ended with \r\n\r
+                            return ParseResult.Incomplete;
+                        }
+
+                        if (_data.Count > 0)
+                        {
+                            //Find the final size of the payload
+                            var payloadSize = 0;
+                            foreach (var dataLine in _data)
+                            {
+                                payloadSize += dataLine.Length;
+                            }
+                            var payload = new byte[payloadSize];
+
+                            //Copy the contents of the data array to a single buffer
+                            var offset = 0;
+                            foreach (var dataLine in _data)
+                            {
+                                dataLine.CopyTo(payload, offset);
+                                offset += dataLine.Length;
+                            }
+
+                            message = new Message(payload, messageType);
+                        }
+                        else
+                        {
+                            //Empty message
+                            message = new Message(Array.Empty<byte>(), messageType);
+                        }
+
+                        consumed = buffer.End;
+                        return ParseResult.Completed;
+                }
+
+                //Peek into next byte. If it is a carriage return byte, then advance to the next state
+                if (reader.Peek() == ByteCR)
+                {
+                    _internalParserState = InternalParseState.ReadEndOfMessage;
+                }
+            }
+            return ParseResult.Incomplete;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private Span<byte> ConvertBufferToSpan(ReadableBuffer buffer)
+        {
+            if (buffer.IsSingleSpan)
+            {
+                return buffer.First.Span;
+            }
+            return buffer.ToArray();
+        }
+
+        public void Reset()
+        {
+            _internalParserState = InternalParseState.ReadMessageType;
+            _data.Clear();
+        }
+
+        private bool StartsWithDataPrefix(ReadOnlySpan<byte> line)
+        {
+            var dataPrefix = Encoding.UTF8.GetBytes("data: ");
```

Comment:
Make this buffer static, it doesn't change :). You can't make the Span static though, since Span can't go on the heap